### PR TITLE
feat: add reading timeout

### PR DESCRIPTION
**Pull request by borisdayma**
Uses timeout during url read (iterate in chunks).

TODO:
* [x] test performance: potentially increase `chunk_size` value if needed

### CODE REVIEW

**Comment by borisdayma** [comment]
Btw linting errors are solved by installing `types-requests`

**Comment by rom1504** [comment]
> Btw linting errors are solved by installing types-requests

I guess add it in test-requirements.txt then ?

File: img2dataset/downloader.py [hunk 1]
```diff
@@ -19,17 +19,27 @@
...
+            timeout=timeout,
+            stream=True,
         )
-        with urllib.request.urlopen(request, timeout=timeout) as r:
-            img_stream = io.BytesIO(r.read())
+        if request.status_code != 200:
+            return key, None, "HTTP status code {}".format(request.status_code)
+        start_time = time.time()
+        img_stream = io.BytesIO()
+        for chunk in request.iter_content(chunk_size=1024):
```
Comment (rom1504) [inline img2dataset/downloader.py:37]: what if the sender takes 10h to send 1024 bytes?

Comment (borisdayma) [reply img2dataset/downloader.py:37]: That's right, the timeout from `requests.get` applies both for opening url and receiving packets when reading.
So the slowest is probably `timeout * 1024`.

The issue is a too low `chunk_size` may slow down the process. My test was already a bit slower so I actually think it needs to be increased to `1024 * 1024`.

It's not perfect but can prevent process potentially hanging forever.

**Comment by borisdayma** [comment]
Completed a test of 500k items:

* main branch -> 12m 51s
* this PR -> 12m 58s, includes 159 items with reading timeout

**Comment by rom1504** [comment]
ok nice!
650 sample/s is a bit slow (I usually see 1300 sample/s with 16 cores 2Gbps and laion400m), what hardware and dataset did you use for testing?

do you have a few samples that failed consistently with the previous code? maybe we could put them in an unit test

**Comment by rom1504** [comment]
it looks pretty good!
I'll run it as well and if I notice nothing wrong, I'll merge

**Comment by rom1504** [comment]
ok so I tried it and I'm afraid it's not fast enough
see comparison there https://wandb.ai/rom1504/img2dataset/reports/request-timeout-pr-153--VmlldzoxNjUxODcx
it's almost 2 times slower now. I believe it's due to higher cpu usage

**Comment by rom1504** [comment]
so I see 2  ways forward here

1. making this fast so it can just be the default thing
2. having 2 downloader implementation and adding a new option to pick it

however I'm really curious if you have an example of failing sample as I didn't have any for downloading 5B samples

**Comment by borisdayma** [comment]
Yes the examples are the ones from the discord but it's true that it's not frequent.
I personally don't think that 2 implementations are convenient so I would use this one only if it goes as fast.

I tested it on TPU VM and because it has tons of CPU cores, maybe this problem doesn't happen.
Unfortunately my home connection is not very performant (I think I have DNS issues due to my Internet provider) so I can't test it there.

I think the issue could potentially be solved with `chunk_size = 1024 * 1024` but if it doesn't fix the issue I would probably just close this PR.

**Comment by rom1504** [comment]
I tried 1024x1024 but it has the same performance characteristics as 1024 https://wandb.ai/rom1504/img2dataset/reports/request-timeout-pr-153--VmlldzoxNjUxODcx 

This hard timeout feature is not easy to implement right. I had the same problem in other similar implementation of downloaders in an other language (scala), and no obvious solution there either.

The problem with such a hard timeout as well is it will clearly decrease the success rate if there's any slow down on the connection, so the value would have to be pretty high

**Comment by borisdayma** [comment]
All right then, last option would be maybe to use [eventlet timeouts](https://eventlet.net/doc/modules/timeout.html) but we may not want other dependencies (and threads).

**Comment by rom1504** [comment]
if you implement something, I'm ok with running my benchmark to see if it's better
if the dependency is not very large and it's not slower I'm ok with it

**Comment by borisdayma** [comment]
Closing this PR.
I'll create a new one with `eventlet`.

**Pull request closed without merging**